### PR TITLE
Add float quickcheck implementations (fixes #45)

### LINKED ISSUE #45: Testing: Add `F32`/`F64` wrappers to test exotic floating point values (INF, -0, NaN)
As done in #36 
Follow up of #43

### PR DESCRIPTION
**Pull request by mattico**
This builds on top of #43 to add Quickcheck helpers for floating point.

I generate a few different types of floats by constructing them from their component parts, using new functions defined on the Float trait. The Add implementation could be refactored to use these new functions for clarity.

### CODE REVIEW

**Comment by mattico** [comment]
Interesting test failure here:
https://travis-ci.org/japaric/rustc-builtins/jobs/153852666#L15056

Slightly different results, possibly a rounding error?

File: src/qc.rs [hunk 1]
```diff
@@ -5,38 +5,40 @@
...
 
 use quickcheck::{Arbitrary, Gen};
 
 use int::LargeInt;
+use float::Float;
 
 // Generates values in the full range of the integer type
 macro_rules! arbitrary {
-    ($TY:ident : $ty:ident) => {
+    ($id:ident: $ty:ty) => {
```
Comment (mattico) [inline src/qc.rs]: I forgot that I left these changes in the commit. I can remove them if you prefer the current style.

File: src/qc.rs [hunk 2]
```diff
@@ -121,3 +123,53 @@ macro_rules! arbitrary_large {
...
+                    let sign: bool = g.gen();
+                    let exponent: <$ty as Float>::Int = 0;
+                    let significand: <$ty as Float>::Int = g.gen();
+                    $id(<$ty>::from_parts(sign, exponent, significand))
+                } else { // NaN variants
+                    let sign: bool = g.gen();
+                    let exponent: <$ty as Float>::Int = g.gen();
+                    let significand: <$ty as Float>::Int = 0;
+                    $id(<$ty>::from_parts(sign, exponent, significand))
+                }
```
Comment (mattico) [inline src/qc.rs]: I'm not entirely convinced by the above method, but the tests get run often enough that I believe that the different combinations will be run many times, and not having any state keeps the method simple.

File: src/qc.rs (hunk 1 above)
Comment (japaric) [inline src/qc.rs]: I would prefer to maintain the status quo for now just to reduce the noise in the diff. I don't have a strong preference for the name of the macro arguments.

**Comment by japaric** [comment]
Interesting that only soft float arm target failed. That means that these intrinsics match the output of hardware instructions (the hard float arm targets pass) but don't match the output of libcompiler-rt.a's intrinsics?

**Comment by homunkulus** [comment]
:umbrella: The latest upstream changes (presumably #67) made this pull request unmergeable. Please resolve the merge conflicts.

**Comment by mattico** [comment]
Rebased

**Comment by mattico** [comment]
I mostly just want to see if this works...
@homunkulus try

**Comment by japaric** [comment]
@mattico ah sorry homu is currently configured to only let me "try". Mostly because changing that default configuration is annoying :P.

@homunkulus try

**Comment by japaric** [comment]
@homunkulus try

(homu was out of sync :X)

**Comment by homunkulus** [comment]
:hourglass: Trying commit 831074e with merge 831074e...

**Comment by homunkulus** [comment]
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/japaric/rustc-builtins/build/1.0.254)

**Comment by mattico** [comment]
Rebased

**Comment by mattico** [comment]
Rebased on #74. Should be mergable now.

**Comment by japaric** [comment]
@homunkulus try

**Comment by homunkulus** [comment]
:hourglass: Trying commit 010d153 with merge 010d153...

**Comment by homunkulus** [comment]
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/japaric/rustc-builtins/build/1.0.261), [status-travis](https://travis-ci.org/japaric/rustc-builtins/builds/164151454)
State: approved= try=True

**Comment by japaric** [comment]
Awesome @mattico!

**Comment by mattico** [comment]
@japaric so is the plan to merge this into rust-lang/rust before doing more work on it?

**Comment by japaric** [comment]
I'm going to try to land this in rust-lang/rust, yeah. But feel free to keep working on it in the mean time :+1:.

**Comment by japaric** [comment]
@mattico could you re-send this PR?

**Pull request merged**